pkg/syncer/storagepool/intended_state.go

```diff
@@ -181,23 +183,36 @@ func (c *SpController) applyIntendedState(ctx context.Context, state *intendedSt
 		if ok && statusErr.Status().Reason == metav1.StatusReasonNotFound {
 			log.Infof("Creating StoragePool instance for %s", state.spName)
 			sp := state.createUnstructuredStoragePool(ctx)
+			// Generation Number of a newly created object is 1. After issuing a write operation the
+			// current thread typically sleeps and wait for server reply. Meanwhile SPWatch can receive
+			// the write event and start to detect if this event was due to an external change. So it is
+			// essential to store expected value in SPNameToGenerationNumberMap before issuing any write
+			// operation to StoragePool to avoid false positive.
+			c.spWatcher.SPNameToGenerationNumberMap[state.spName] = 1
```

Comment:
I haven't looked at the overall design, but in case CSI syncer restarts, this in-memory number will be reset. Does this functionality cover this scenario?

Reply:
Yes, we support syncer restart. What we are doing is comparing the expected generation number with the number received from storage pool instance. If number received is greater than expected number we do a reconciliation of all storage pools which is a non-destructive operation. Just after the restart we do not have the data of actual generation number so when the first update occurs we may trigger a full reconciliation which fills SPNameToGenerationNumberMap with correct info for all existing storage pool (also fills up other empty caches just after restart which is a win-win). For any subsequent updates we do not trigger full reconciliation unless it was made by external user.